src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs

```diff
@@ -497,7 +497,11 @@ public override long Position
         // _strategy can be null only when ctor has thrown
         protected override void Dispose(bool disposing) => _strategy?.DisposeInternal(disposing);
 
-        public override ValueTask DisposeAsync() => _strategy.DisposeAsync();
+        public override ValueTask DisposeAsync()
+        {
+            GC.SuppressFinalize(this);
+            return _strategy.DisposeAsync();
```

Comment:
The pattern @bartonjs landed on for this is outlined in:
https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#the-disposeasync-method
It's complicated by how disposal is now split between different layers of this strategy pattern, but I think here the _strategy.DisposeAsync is essentially the DisposeAsyncCore method, which would lead to this being:
```C#
await _strategy.DisposeAsync().ConfigureAwait(false);
_strategy.DisposeInternal(false);
GC.SuppressFinalize(this);
```
What you have above might be fine, but it does result in a difference between whether finalization is suppressed if disposal throws an exception, as Stream.Dispose will suppress finalization after calling Dispose(true) whereas what's in the PR will suppress before.

Reply:
Since FileStream isn't sealed, it should stick with the pattern, and not inline the call to Dispose(false), since that could be overridden.

```C#
await _strategy.DisposeAsync().ConfigureAwait(false);
Dispose(false);
GC.SuppressFinalize(this);
```

Reply:
> Since FileStream isn't sealed

Gaaaaaahhhhhhhhhhhh....

We need to change our guidance here :)

Reply:
@stephentoub @bartonjs thank you both for the feedback and guidance!